yamux/src/connection/stream.rs

```diff
@@ -200,33 +247,167 @@ impl Stream {
-            ready!(self
-                .sender
-                .poll_ready(cx)
-                .map_err(|_| self.write_zero_err())?);
-
-            shared.window += credit;
-            drop(shared);
-
-            let mut frame = Frame::window_update(self.id, credit).right();
-            self.add_flag(frame.header_mut());
-            let cmd = StreamCommand::SendFrame(frame);
-            self.sender
-                .start_send(cmd)
-                .map_err(|_| self.write_zero_err())?;
+        let bytes_received = shared.max_receive_window_size - shared.current_receive_window_size;
+        let mut next_window_update =
+            bytes_received.saturating_sub(shared.buffer.len().try_into().unwrap_or(u32::MAX));
+
+        // Don't send an update in case half or more of the window is still available to the sender.
+        if next_window_update < shared.max_receive_window_size / 2 {
+            return None;
         }
 
-        Poll::Ready(Ok(()))
+        log::trace!(
+            "received {} mb in {} seconds ({} mbit/s)",
+            next_window_update as f64 / 1024.0 / 1024.0,
+            self.last_window_update.elapsed().as_secs_f64(),
+            next_window_update as f64 / 1024.0 / 1024.0 * 8.0
+                / self.last_window_update.elapsed().as_secs_f64()
+        );
+
+        // Auto-tuning `max_receive_window_size`
+        //
+        // The ideal `max_receive_window_size` is equal to the bandwidth-delay-product (BDP), thus
+        // allowing the remote sender to exhaust the entire available bandwidth on a single stream.
+        // Choosing `max_receive_window_size` too small prevents the remote sender from exhausting
+        // the available bandwidth. Choosing `max_receive_window_size` to large is wasteful and
+        // delays backpressure from the receiver to the sender on the stream.
+        //
+        // In case the remote sender has exhausted half or more of its credit in less than 2
+        // round-trips, try to double `max_receive_window_size`.
+        //
+        // For simplicity `max_receive_window_size` is never decreased.
+        //
+        // This implementation is heavily influenced by QUIC. See document below for rational on the
+        // above strategy.
+        //
+        // https://docs.google.com/document/d/1F2YfdDXKpy20WVKJueEf4abn_LVZHhMUMS5gX6Pgjl4/edit?usp=sharing
+        if self
+            .rtt
+            .get()
+            .map(|rtt| self.last_window_update.elapsed() < rtt * 2)
+            .unwrap_or(false)
+        {
+            let mut accumulated_max_stream_windows = self.accumulated_max_stream_windows.lock();
+
+            // Ideally one can just double it:
+            let mut new_max = shared.max_receive_window_size.saturating_mul(2);
+
+            // Then one has to consider the configured stream limit:
+            new_max = cmp::min(
+                new_max,
+                self.config.max_stream_receive_window.unwrap_or(u32::MAX),
+            );
```

Comment:
Why do you prefer shadowing over re-assignment? Shadowing bears the risk of creating two variables, with a typo in one. Re-assignment makes that impossible. What I am doing here semantically is re-assignment.

Reply:
> Shadowing bears the risk of creating two variables, with a typo in one.

I think in most cases, that should yield an "unused variable" lint somewhere.

Reply:
> > If we can remove `mut`, I know that this variable is never re-assigned which IMO is easier to read.
> 
> But you still have to scan the entire function to check whether it is ever shadowed. I argue that this is the same. Only that the `mut` keyword makes it explicit, whereas shadowing is implicit.

With the difference that shadowing is always unconditional because a condition always introduce a nested scope (you can't shadow a variable from within an `if`).  I'd recommend following the rule of least power:

`mut` is more powerful than shadowing, hence its use should be justified. If you don't need `mut`, don't use it! A reader of this code following this rule will otherwise see the `mut`, switch their brain to "careful thinking, we are mutating state here"-mode only to find out that it could have been done without `mut` :)